core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,874 @@
+   */
+  final def zipWithIndex: Chunk[(A, Int)] = zipWithIndex0(0)
+
+  protected[zio] def apply(n: Int): A
+  protected[zio] def foreach(f: A => Unit): Unit
+
+  //noinspection AccessorLikeMethodIsUnit
+  protected[zio] def toArray[A1 >: A](n: Int, dest: Array[A1]): Unit
+}
+
+object Chunk {
+
+  /**
+   * Returns a chunk from a number of values.
+   */
+  final def apply[A](as: A*): Chunk[A] = fromIterable(as)
+
+  /**
+   * Returns the empty chunk.
+   */
+  final val empty: Chunk[Nothing] = Empty
+
+  /**
+   * Returns a chunk backed by an array.
+   */
+  final def fromArray[@specialized A](array: Array[A]): Chunk[A] = Arr(array)
+
+  /**
+   * Returns a chunk backed by an iterable.
+   */
+  final def fromIterable[A](it: Iterable[A]): Chunk[A] =
+    if (it.size <= 0) Empty
+    else if (it.size == 1) Singleton(it.head)
+    else {
+      it match {
+        case l: Vector[A] => VectorChunk(l)
+        case _ =>
+          val first = it.head
+
+          implicit val A: ClassTag[A] = Tags.fromValue(first)
+
+          fromArray(it.toArray)
+      }
+    }
+
+  /**
+   * Returns a singleton chunk.
+   */
+  final def point[@specialized A](a: A): Chunk[A] = Singleton(a)
+
+  /**
+   * Returns the `ClassTag` for the element type of the chunk.
+   */
+  private final def classTagOf[A](chunk: Chunk[A]): ClassTag[A] = chunk match {
+    case x: Arr[A]         => x.classTag
+    case x: Concat[A]      => x.classTag
+    case Empty             => classTag[java.lang.Object].asInstanceOf[ClassTag[A]]
+    case x: Singleton[A]   => x.classTag
+    case x: Slice[A]       => x.classTag
+    case x: VectorChunk[A] => x.classTag
+  }
+
+  private case class Arr[@specialized A](private val array: Array[A]) extends Chunk[A] {
+    implicit lazy val classTag: ClassTag[A] = ClassTag(array.getClass.getComponentType)
+
+    override def collect[@specialized B](p: PartialFunction[A, B]): Chunk[B] = {
+      val self = array
+      val len  = self.length
+      var dest = null.asInstanceOf[Array[B]]
+
+      var i = 0
+      var j = 0
+      while (i < len) {
+        val a = self(i)
+
+        if (p.isDefinedAt(a)) {
+          val b = p(a)
+
+          if (dest == null) {
+            implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+
+            dest = Array.ofDim[B](len)
+          }
+
+          dest(j) = b
+
+          j += 1
+        }
+
+        i += 1
+      }
+
+      if (dest == null) Chunk.Empty
+      else Chunk.Slice(Chunk.Arr(dest), 0, j)
+    }
+
+    override def dropWhile(f: A => Boolean): Chunk[A] = {
+      val self = array
+      val len  = self.length
+
+      var i = 0
+      while (i < len && f(self(i))) {
+        i += 1
+      }
+
+      if (i == len) Chunk.empty
+      else if (i == 0) this
+      else Chunk.Slice(this, i, len - i)
+    }
+
+    override def filter(f: A => Boolean): Chunk[A] = {
+      val self = array
+      val len  = self.length
+      val dest = Array.ofDim[A](len)
```

Comment:
Might be useful to do the `null` trick here too

Reply:
`filter` actually needs to be reworked to use a strategy in which the array size doubles as needed, so leaving this to a later PR